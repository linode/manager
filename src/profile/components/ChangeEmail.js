import React, { PropTypes, Component } from 'react';

import { Card, CardHeader } from 'linode-components/cards';
import {
  Form,
  FormGroup,
  FormGroupError,
  FormSummary,
  SubmitButton,
  Input,
} from 'linode-components/forms';

import { profile } from '~/api';
import { dispatchOrStoreErrors } from '~/api/util';


export default class ChangeEmail extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,
      errors: {},
      email: props.email,
    };
  }

  onSubmit = () => {
    const { dispatch } = this.props;
    const { email } = this.state;

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => profile.put({ email }),
    ]));
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  render() {
    const { loading, errors, email } = this.state;

    return (
      <Card header={<CardHeader title="Change email" />}>
        <Form onSubmit={this.onSubmit}>
          <FormGroup className="row" errors={errors} name="email">
            <label htmlFor="email" className="col-sm-2 col-form-label">Email</label>
            <div className="col-sm-10">
              <Input
                id="email"
                type="email"
                name="email"
                onChange={this.onChange}
                value={email}
              />
              <FormGroupError errors={errors} name="email" />
            </div>
          </FormGroup>
          <FormGroup className="row">
            <div className="offset-sm-2 col-sm-10">
              <SubmitButton disabled={loading} />
              <FormSummary errors={errors} success="Email saved." />
            </div>
          </FormGroup>
        </Form>
      </Card>
    );
  }
}

ChangeEmail.propTypes = {
  email: PropTypes.string.isRequired,
  dispatch: PropTypes.func.isRequired,
};
