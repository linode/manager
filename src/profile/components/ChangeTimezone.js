import moment from 'moment-timezone';
import React, { PropTypes, Component } from 'react';

import { Card, CardHeader } from 'linode-components/cards';
import {
  Form,
  FormGroup,
  FormSummary,
  Select,
  SubmitButton,
} from 'linode-components/forms';

import { profile } from '~/api';
import { dispatchOrStoreErrors } from '~/api/util';
import { setStorage } from '~/storage';


const timezones = moment.tz.names();

export default class ChangeTimezone extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loading: false,
      errors: {},
      timezone: props.timezone,
    };
  }

  onSubmit = () => {
    const { dispatch } = this.props;
    const { timezone } = this.state;

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => profile.put({ timezone }),
      () => setStorage('profile/timezone', timezone),
    ]));
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  render() {
    const { loading, errors, timezone } = this.state;

    return (
      <Card header={<CardHeader title="Change timezone" />}>
        <Form onSubmit={this.onSubmit}>
          <FormGroup className="row">
            <label htmlFor="timezone" className="col-sm-2 col-form-label">Timezone</label>
            <div className="col-sm-10">
              <Select
                id="timezone"
                name="timezone"
                onChange={this.onChange}
                value={timezone}
                options={timezones.map(zone => ({ value: zone, label: zone }))}
              />
              <div>
                <small className="text-muted">
                  This timezone setting is unique to this app. Any timezone setting in the classic
                  Manager will not be reflected here.
                </small>
              </div>
            </div>
          </FormGroup>
          <FormGroup className="row">
            <div className="offset-sm-2 col-sm-10">
              <SubmitButton disabled={loading} />
              <FormSummary errors={errors} success="Timezone saved." />
            </div>
          </FormGroup>
        </Form>
      </Card>
    );
  }
}

ChangeTimezone.propTypes = {
  timezone: PropTypes.string.isRequired,
  dispatch: PropTypes.func.isRequired,
};
