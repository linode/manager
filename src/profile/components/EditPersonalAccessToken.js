import React, { PropTypes, Component } from 'react';

import { Input, ModalFormGroup } from 'linode-components/forms';
import { FormModalBody } from 'linode-components/modals';

import { hideModal, showModal } from '~/actions/modal';
import { tokens } from '~/api';
import { dispatchOrStoreErrors } from '~/api/util';


export default class EditPersonalAccessToken extends Component {
  static title = 'Edit Personal Access Token'

  static trigger(dispatch, token) {
    dispatch(showModal(EditPersonalAccessToken.title, (
      <EditPersonalAccessToken
        id={token.id}
        label={token.label}
        dispatch={dispatch}
        close={() => dispatch(hideModal())}
      />
    )));
  }

  constructor(props) {
    super(props);

    this.state = {
      errors: {},
      label: props.label,
    };
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  onSubmit = () => {
    const { dispatch, id, close } = this.props;
    const { label } = this.state;

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => tokens.put({ label }, id),
      close,
    ]));
  }

  render() {
    const { errors, label } = this.state;

    return (
      <FormModalBody
        onSubmit={this.onSubmit}
        onCancel={this.props.close}
        errors={errors}
        analytics={{ title: EditPersonalAccessToken.title }}
      >
        <ModalFormGroup id="label" label="Label" apiKey="label" errors={errors}>
          <Input
            name="label"
            id="label"
            placeholder="My token"
            value={label}
            onChange={this.onChange}
          />
        </ModalFormGroup>
      </FormModalBody>
    );
  }
}

EditPersonalAccessToken.propTypes = {
  dispatch: PropTypes.func.isRequired,
  close: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
};
