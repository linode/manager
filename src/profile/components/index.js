export { default as ChangeTimezone } from './ChangeTimezone';
export { default as ChangeEmail } from './ChangeEmail';
export { default as ChangePassword } from './ChangePassword';
export { default as TwoFactor } from './TwoFactor';
