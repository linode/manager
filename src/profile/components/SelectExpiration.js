import React, { PropTypes, Component } from 'react';
import moment from 'moment';

import { Select } from '~/components/form';

export default class SelectExpiration extends Component {
  static map(value) {
    return value === '0' ? undefined : (
      moment()
        .add(value, 'days')
        .utc()
        .format('YYYY-MM-DDTHH:mm:ss')
        .toString());
  }

  render() {
    return (
      <Select
        id={this.props.id}
        onChange={this.props.onChange}
        value={this.props.value}
        name={this.props.name}
      >
        <option value="0">Never</option>
        <option value="30">In 1 month</option>
        <option value="90">In 3 months</option>
        <option value="180">In 6 months</option>
      </Select>
    );
  }
}

SelectExpiration.propTypes = {
  value: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  name: PropTypes.string.isRequired,
};
