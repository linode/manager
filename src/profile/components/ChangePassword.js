import React, { Component, PropTypes } from 'react';

import { Card, CardHeader } from 'linode-components/cards';
import {
  Form,
  FormGroup,
  FormGroupError,
  FormSummary,
  PasswordInput,
  SubmitButton,
} from 'linode-components/forms';

import { setPassword } from '~/api/profile';
import { dispatchOrStoreErrors } from '~/api/util';

import SelectExpiration from '../components/SelectExpiration';


export default class ChangePassword extends Component {
  constructor() {
    super();

    this.state = {
      password: '',
      expires: '0',
      errors: {},
      loading: false,
    };
  }

  onSubmit = () => {
    const { dispatch } = this.props;
    const { password, expires } = this.state;

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => setPassword(password, SelectExpiration.map(expires)),
      () => this.setState({ password: '', expires: '0' }),
    ]));
  }

  onChange = ({ target: { value, name } }) => this.setState({ [name]: value })

  render() {
    const { password, expires, errors, loading } = this.state;

    return (
      <Card header={<CardHeader title="Change password" />}>
        <Form onSubmit={this.onSubmit}>
          <FormGroup className="row" errors={errors} name="password">
            <label htmlFor="password" className="col-sm-2 col-form-label">New password</label>
            <div className="col-sm-10">
              <PasswordInput
                name="password"
                id="password"
                value={password}
                onChange={this.onChange}
              />
              <FormGroupError errors={errors} name="password" />
            </div>
          </FormGroup>
          <FormGroup className="row" errors={errors} name="expires">
            <label className="col-sm-2 col-form-label">Expires</label>
            <div className="col-sm-10">
              <SelectExpiration
                id="expires"
                name="expires"
                value={expires}
                onChange={this.onChange}
              />
              <FormGroupError errors={errors} name="expires" />
            </div>
          </FormGroup>
          <FormGroup className="row">
            <div className="col-sm-10 offset-sm-2">
              <SubmitButton disabled={loading} />
              <FormSummary errors={errors} success="Password changed." />
            </div>
          </FormGroup>
        </Form>
      </Card>
    );
  }
}

ChangePassword.propTypes = {
  dispatch: PropTypes.func.isRequired,
};
