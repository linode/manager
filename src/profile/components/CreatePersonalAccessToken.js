import _ from 'lodash';
import React, { PropTypes, Component } from 'react';

import { Input, ModalFormGroup, Select } from 'linode-components/forms';
import { FormModalBody } from 'linode-components/modals';

import { showModal, hideModal } from '~/actions/modal';
import { tokens } from '~/api';
import { dispatchOrStoreErrors } from '~/api/util';
import { OAUTH_SUBSCOPES, OAUTH_SCOPES } from '~/constants';

import SelectExpiration from './SelectExpiration';
import { formatScope } from '../utilities';


export function renderSecret(label, verb, secret, close) {
  const title = [...label.split(' '), verb].map(_.capitalize).join(' ');
  return showModal(title,
    <FormModalBody
      onSubmit={close}
      buttonText="I understand"
      noCancel
      analytics={{ title, action: 'info' }}
    >
      <div>
        <p>
          Your {label} has been {verb}. Store this secret. It won't be shown again.
        </p>
        <div className="alert alert-warning" id="secret">{secret}</div>
      </div>
    </FormModalBody>
  );
}

export default class CreatePersonalAccessToken extends Component {
  static title = 'Create a Personal Access Token'

  static trigger(dispatch) {
    return dispatch(showModal(CreatePersonalAccessToken.title, (
      <CreatePersonalAccessToken
        dispatch={dispatch}
        close={() => dispatch(hideModal())}
      />
    )));
  }

  constructor() {
    super();

    this.state = {
      ...OAUTH_SCOPES.reduce((object, scope) => ({
        ...object,
        [scope]: OAUTH_SUBSCOPES[OAUTH_SUBSCOPES.length - 1],
      }), {}),
      errors: {},
      label: '',
      expiry: null,
      loading: false,
    };
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  onSubmit = () => {
    const { dispatch } = this.props;
    const { label, expiry } = this.state;

    const scopes = OAUTH_SCOPES.reduce((scopes, scope) => {
      const level = this.state[scope];
      if (OAUTH_SUBSCOPES.indexOf(level) === -1) {
        return scopes;
      }

      return [...scopes, `${scope}:${level}`];
    }, []).join(',') || ''; // '' allows no scopes. undefined copies current scopes.

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => tokens.post({ label, scopes, expiry: SelectExpiration.map(expiry) }),
      ({ token }) => renderSecret('personal access token', 'created', token, this.props.close),
    ]));
  }

  renderScopeFormGroup = (scope) => {
    return (
      <ModalFormGroup
        id={scope}
        label={formatScope(scope)}
        apiKey={scope}
        errors={this.state.errors}
        key={scope}
      >
        <Select
          name={scope}
          id={scope}
          value={this.state[scope]}
          onChange={this.onChange}
          options={['No Access', ...OAUTH_SUBSCOPES].reverse().map(value => ({
            value,
            label: `${_.capitalize(value)}${value === 'delete' ? ' (All Access)' : ''}`,
          }))}
        />
      </ModalFormGroup>
    );
  }

  render() {
    const { close } = this.props;
    const { errors, label, expiry } = this.state;

    return (
      <FormModalBody
        onSubmit={this.onSubmit}
        onCancel={close}
        buttonText="Create"
        buttonDisabledText="Creating"
        errors={errors}
        analytics={{ title: CreatePersonalAccessToken.title, action: 'add' }}
      >
        <div>
          <ModalFormGroup id="label" label="Label" apiKey="label" errors={errors}>
            <Input
              name="label"
              id="label"
              placeholder="My client"
              value={label}
              onChange={this.onChange}
            />
          </ModalFormGroup>
          <ModalFormGroup id="expiry" label="Expires" apiKey="expiry" errors={errors}>
            <SelectExpiration
              value={expiry}
              name="expiry"
              id="expiry"
              onChange={this.onChange}
            />
          </ModalFormGroup>
          <h3 className="sub-header">Access</h3>
          <p><small>
            Set a maximum (cascading) access level for every OAuth scope to restrict this token's
            abilities. e.g. Modify access also gives access to View.
          </small></p>
          {OAUTH_SCOPES.map(this.renderScopeFormGroup)}
        </div>
      </FormModalBody>
    );
  }
}

CreatePersonalAccessToken.propTypes = {
  dispatch: PropTypes.func.isRequired,
  close: PropTypes.func.isRequired,
};
