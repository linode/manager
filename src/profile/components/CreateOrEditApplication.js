import React, { PropTypes, Component } from 'react';

import { Input, ModalFormGroup } from 'linode-components/forms';
import { FormModalBody } from 'linode-components/modals';

import { hideModal, showModal } from '~/actions/modal';
import { clients } from '~/api';
import { updateClientThumbnail } from '~/api/clients';
import { dispatchOrStoreErrors } from '~/api/util';
import { MAX_UPLOAD_SIZE_MB } from '~/constants';

import { renderSecret } from './CreatePersonalAccessToken';


export default class CreateOrEditApplication extends Component {
  static trigger(dispatch, client = {}) {
    const title = client ? 'Create an OAuth Client' : 'Edit OAuth Client';

    return dispatch(showModal(title, (
      <CreateOrEditApplication
        dispatch={dispatch}
        close={() => dispatch(hideModal())}
        title={title}
        {...client}
      />
    )));
  }

  constructor(props) {
    super(props);

    this.state = {
      label: props.label || '',
      redirect: props.redirect_uri || '',
      thumbnail: props.thumbnail || '',
      errors: {},
    };
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  onSubmit = () => {
    const { dispatch, id, close } = this.props;
    const { label, redirect, thumbnail } = this.state;

    const data = { label, redirect_uri: redirect };
    const idsPath = [id].filter(Boolean);

    return dispatch(dispatchOrStoreErrors.call(this, [
      () => clients[id ? 'put' : 'post'](data, ...idsPath),
      ({ id }) => {
        if (thumbnail) {
          if ((thumbnail.size / (1024 * 1024)) < MAX_UPLOAD_SIZE_MB) {
            return updateClientThumbnail(id, thumbnail);
          }

          // eslint-disable-next-line no-throw-literal
          throw { json: () => Promise.resolve({
            errors: [{
              field: 'thumbnail',
              reason: `File size must be under ${MAX_UPLOAD_SIZE_MB} MB`,
            }],
          }) };
        }
      },
      ({ secret }) => id ? close() :
        renderSecret('client', 'created', secret, close),
    ]));
  }

  render() {
    const { close, title, id } = this.props;
    const { errors, label, redirect } = this.state;

    return (
      <FormModalBody
        onSubmit={this.onSubmit}
        onCancel={close}
        errors={errors}
        analytics={{ title, action: id ? 'edit' : 'add' }}
      >
        <div>
          <ModalFormGroup id="label" label="Label" apiKey="label" errors={errors}>
            <Input
              name="label"
              id="label"
              placeholder="My client"
              value={label}
              onChange={this.onChange}
            />
          </ModalFormGroup>
          <ModalFormGroup id="redirect" label="Redirect URI" apiKey="redirect_uri" errors={errors}>
            <Input
              name="redirect"
              id="redirect"
              placeholder="http://localhost:3000/oauth/callback"
              value={redirect}
              onChange={this.onChange}
            />
          </ModalFormGroup>
          <ModalFormGroup id="thumbnail" label="Thumbnail" apiKey="thumbnail" errors={errors}>
            <Input
              name="thumbnail"
              id="thumbnail"
              type="file"
              onChange={(e) => this.setState({ thumbnail: e.target.files[0] })}
            />
          </ModalFormGroup>
        </div>
      </FormModalBody>
    );
  }
}

CreateOrEditApplication.propTypes = {
  dispatch: PropTypes.func.isRequired,
  close: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  label: PropTypes.string,
  id: PropTypes.string,
  redirect_uri: PropTypes.string,
  thumbnail: PropTypes.string,
};
