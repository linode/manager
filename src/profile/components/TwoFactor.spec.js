import { shallow } from 'enzyme';
import React from 'react';
import sinon from 'sinon';

import { TwoFactor } from '~/profile/components';

import { expectDispatchOrStoreErrors, expectRequest } from '~/test.helpers';


describe('profile/components/TwoFactor', () => {
  const sandbox = sinon.sandbox.create();

  const dispatch = sandbox.stub();

  it('disable two factor called', async () => {
    const page = shallow(
      <TwoFactor
        dispatch={dispatch}
        tfaEnabled
      />
    );

    const button = page.find('SubmitButton').at(0);
    expect(button.props().children).toBe('Disable');

    dispatch.reset();
    await page.find('Form').props().onSubmit({ preventDefault() { } });
    expect(dispatch.callCount).toBe(1);
    await expectDispatchOrStoreErrors(dispatch.firstCall.args[0], [
      ([fn]) => expectRequest(fn, '/profile/tfa-disable', {
        method: 'POST',
      }),
    ], 1);
  });

  it('enable two factor called', async () => {
    const page = shallow(
      <TwoFactor
        dispatch={dispatch}
        tfaEnabled={false}
      />
    );

    const button = page.find('SubmitButton').at(0);
    expect(button.props().children).toBe('Enable');

    dispatch.reset();
    await page.find('Form').props().onSubmit({ preventDefault() { } });
    expect(dispatch.callCount).toBe(1);
    await expectDispatchOrStoreErrors(dispatch.firstCall.args[0], [
      ([fn]) => expectRequest(fn, '/profile/tfa-enable', {
        method: 'POST',
      }),
    ], 1, [{ secret: '' }]);
  });
});
