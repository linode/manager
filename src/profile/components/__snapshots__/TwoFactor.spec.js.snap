// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`profile/components/TwoFactor should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TwoFactor
    dispatch={[Function]}
    tfaEnabled={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        null,
        <Form
          analytics={
            Object {
              "title": "Change Two-Factor Authentication",
            }
          }
          className=""
          noSubmitEvent={false}
          onSubmit={[Function]}
        >
          <p>
            Two-factor authentication (TFA) is currently 
            <strong>
              enabled
            </strong>
            .
          </p>
          <SubmitButton
            className="btn-default"
            disabled={false}
            disabledChildren="Disabling"
          >
            Disable
          </SubmitButton>
          <FormSummary
            className=""
            errors={Object {}}
          />
        </Form>,
      ],
      "className": "",
      "header": <CardHeader
        title="Change Two-Factor Authentication"
      />,
    },
    "ref": null,
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "analytics": Object {
            "title": "Change Two-Factor Authentication",
          },
          "children": Array [
            <p>
              Two-factor authentication (TFA) is currently 
              <strong>
                enabled
              </strong>
              .
            </p>,
            <SubmitButton
              className="btn-default"
              disabled={false}
              disabledChildren="Disabling"
            >
              Disable
            </SubmitButton>,
            <FormSummary
              className=""
              errors={Object {}}
            />,
          ],
          "className": "",
          "noSubmitEvent": false,
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Two-factor authentication (TFA) is currently ",
                <strong>
                  enabled
                </strong>,
                ".",
              ],
            },
            "ref": null,
            "rendered": Array [
              "Two-factor authentication (TFA) is currently ",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "enabled",
                },
                "ref": null,
                "rendered": "enabled",
                "type": "strong",
              },
              ".",
            ],
            "type": "p",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Disable",
              "className": "btn-default",
              "disabled": false,
              "disabledChildren": "Disabling",
            },
            "ref": null,
            "rendered": "Disable",
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "",
              "errors": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          null,
          <Form
            analytics={
              Object {
                "title": "Change Two-Factor Authentication",
              }
            }
            className=""
            noSubmitEvent={false}
            onSubmit={[Function]}
          >
            <p>
              Two-factor authentication (TFA) is currently 
              <strong>
                enabled
              </strong>
              .
            </p>
            <SubmitButton
              className="btn-default"
              disabled={false}
              disabledChildren="Disabling"
            >
              Disable
            </SubmitButton>
            <FormSummary
              className=""
              errors={Object {}}
            />
          </Form>,
        ],
        "className": "",
        "header": <CardHeader
          title="Change Two-Factor Authentication"
        />,
      },
      "ref": null,
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "analytics": Object {
              "title": "Change Two-Factor Authentication",
            },
            "children": Array [
              <p>
                Two-factor authentication (TFA) is currently 
                <strong>
                  enabled
                </strong>
                .
              </p>,
              <SubmitButton
                className="btn-default"
                disabled={false}
                disabledChildren="Disabling"
              >
                Disable
              </SubmitButton>,
              <FormSummary
                className=""
                errors={Object {}}
              />,
            ],
            "className": "",
            "noSubmitEvent": false,
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Two-factor authentication (TFA) is currently ",
                  <strong>
                    enabled
                  </strong>,
                  ".",
                ],
              },
              "ref": null,
              "rendered": Array [
                "Two-factor authentication (TFA) is currently ",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "enabled",
                  },
                  "ref": null,
                  "rendered": "enabled",
                  "type": "strong",
                },
                ".",
              ],
              "type": "p",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": "Disable",
                "className": "btn-default",
                "disabled": false,
                "disabledChildren": "Disabling",
              },
              "ref": null,
              "rendered": "Disable",
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "",
                "errors": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
